storage/innobase/xtrabackup/src/keyring_plugins.cc

```diff
@@ -236,24 +251,69 @@ argc and argv.
 bool
 xb_keyring_init_for_stats(int argc, char **argv)
 {
-	mysql_optional_plugins[0] = 0;
-	mysql_mandatory_plugins[0] = 0;
+	char *plugin_load = NULL;
 
-	if (opt_xtra_plugin_dir != NULL) {
-		strcpy(opt_plugin_dir, opt_xtra_plugin_dir);
-	} else {
-		strcpy(opt_plugin_dir, PLUGINDIR);
+	my_option keyring_options[] = {
+		{"plugin-load", 0, "", &plugin_load, &plugin_load, 0, GET_STR,
+		 REQUIRED_ARG, 0, 0, 0, 0, 0, 0},
+		{0, 0, 0, 0, 0, 0, GET_NO_ARG, NO_ARG, 0, 0, 0, 0, 0, 0}
+	};
+
+	if (handle_options(&argc, &argv, keyring_options, get_one_option)) {
+		return(false);
 	}
 
+	if (plugin_load != NULL) {
+		opt_plugin_load_list_ptr->push_back(
+			new i_string(my_strdup(
+				PSI_NOT_INSTRUMENTED, plugin_load,
+				MYF(MY_FAE))));
+	}
+
+	init_plugins(argc, argv);
+
+	return(true);
+}
+
+/** Initialize keyring plugin for stats mode. Configuration is read from
+argc and argv, server uuid and plugin name is read from backup-my.cnf.
+@param[in, out]	argc	Command line options (count)
+@param[in, out]	argv	Command line options (values)
+@return true if success */
+bool
+xb_keyring_init_for_prepare(int argc, char **argv)
+{
+	char *uuid = NULL;
 	char *plugin_load = NULL;
+	const char *groups[] = {"mysqld", NULL};
 
 	my_option keyring_options[] = {
+		{"server-uuid", 0, "", &uuid, &uuid, 0, GET_STR,
+		 REQUIRED_ARG, 0, 0, 0, 0, 0, 0},
 		{"plugin-load", 0, "", &plugin_load, &plugin_load, 0, GET_STR,
 		 REQUIRED_ARG, 0, 0, 0, 0, 0, 0},
 		{0, 0, 0, 0, 0, 0, GET_NO_ARG, NO_ARG, 0, 0, 0, 0, 0, 0}
 	};
 
-	if (handle_options(&argc, &argv, keyring_options, get_one_option)) {
+	char *exename = (char *) "xtrabackup";
```

Comment:
Does it really have to be a non-const string? This is definitely at least a warning, and dangerous - maybe create a modifiable instance with `strdup`?

Reply:
`handle_options` should not really modify argv[0]